src/lobpcg/algorithm.rs

```diff
@@ -0,0 +1,658 @@
+    y: ArrayView2<A>,
+) {
+    let gram_yv = y.t().dot(&v);
+
+    /*
+    let u = gram_yv
+        .columns()
+        .into_iter()
+        .map(|x| {
+            let res = cholesky_yy.solve_triangular(&x, UPLO::Upper)?;
+
+            res.to_vec()
+        })
+        .flatten()
+        .collect::<Vec<A>>();
+
+    let rows = gram_yv.len_of(Axis(0));
+    let u = Array2::from_shape_vec((rows, u.len() / rows), u).unwrap();*/
+    let u = cholesky_yy.solve_triangular(&gram_yv, UPLO::Lower).unwrap();
+
+    v -= &(y.dot(&u));
+}
+
+/// Orthonormalize `V` with Cholesky factorization
+///
+/// This also returns the matrix `R` of the `QR` problem
+fn orthonormalize<T: NdFloat>(v: Array2<T>) -> Result<(Array2<T>, Array2<T>)> {
+    let gram_vv = v.t().dot(&v);
+    let gram_vv_fac = gram_vv.cholesky()?;
+
+    //assert_abs_diff_eq!(
+    //    &gram_vv,
+    //    &gram_vv_fac.dot(&gram_vv_fac.t()),
+    //    epsilon=NumCast::from(1e-5).unwrap(),
+    //);
+
+    let v_t = v.reversed_axes();
+    let u = gram_vv_fac
+        .solve_triangular(&v_t, UPLO::Lower)?
+        .reversed_axes();
+
+    Ok((u, gram_vv_fac))
+}
+
+/// Eigenvalue solver for large symmetric positive definite (SPD) eigenproblems
+///
+/// # Arguments
+/// * `a` - An operator defining the problem, usually a sparse (sometimes also dense) matrix
+/// multiplication. Also called the "stiffness matrix".
+/// * `x` - Initial approximation of the k eigenvectors. If `a` has shape=(n,n), then `x` should
+/// have shape=(n,k).
+/// * `m` - Preconditioner to `a`, by default the identity matrix. Should approximate the inverse
+/// of `a`.
+/// * `y` - Constraints of (n,size_y), iterations are performed in the orthogonal complement of the
+/// column-space of `y`. It must be full rank.
+/// * `tol` - The tolerance values defines at which point the solver stops the optimization. The approximation
+/// of a eigenvalue stops when then l2-norm of the residual is below this threshold.
+/// * `maxiter` - The maximal number of iterations
+/// * `order` - Whether to solve for the largest or lowest eigenvalues
+///
+/// The function returns an `LobpcgResult` with the eigenvalue/eigenvector and achieved residual norm
+/// for it. All iterations are tracked and the optimal solution returned. In case of an error a
+/// special variant `LobpcgResult::NotConverged` additionally carries the error. This can happen when
+/// the precision of the matrix is too low (switch then from `f32` to `f64` for example).
+pub fn lobpcg<
+    A: Float + NdFloat + Sum + ScalarOperand + PartialOrd + Default,
+    F: Fn(ArrayView2<A>) -> Array2<A>,
+    G: Fn(ArrayViewMut2<A>),
+>(
+    a: F,
+    mut x: Array2<A>,
+    m: G,
+    y: Option<Array2<A>>,
+    tol: f32,
+    maxiter: usize,
+    order: Order,
+) -> LobpcgResult<A> {
+    // the initital approximation should be maximal square
+    // n is the dimensionality of the problem
+    let (n, size_x) = (x.nrows(), x.ncols());
+    assert!(size_x <= n);
```

Comment:
We're checking that `x` is a "thin" matrix (rows >= cols) here right? Then we should return `LinalgError::NotThin` if the dimensions are wrong instead of panicking, since that's what the rest of the crate does.